scripts/check.sh

```diff
@@ -165,17 +165,30 @@ fi
 
 # When travis clones a repo for building, it uses a shallow clone. After the
 # first commit on a non-master branch, TRAVIS_COMMIT_RANGE is not set and
-# START_REVISION is "master" instead of a range.
+# START_REVISION is "origin/master" because a range wasn't passed in.
+if [[ "${START_REVISION}" == *..* ]]; then
+  RANGE_START="${START_REVISION/..*/}"
+  RANGE_END="${START_REVISION/*../}"
 
-# If needed, check if we have access to master and add it to the repo.
-if [[ "${START_REVISION}" == "origin/master" ]]; then
+  # Figure out if we have access to origin/master. If not add it to the repo.
   if ! git rev-parse origin/master >& /dev/null; then
     git remote set-branches --add origin master
     git fetch origin
   fi
-fi
 
-START_SHA=$(git rev-parse "${START_REVISION}")
+  NEW_RANGE_START=$(git merge-base origin/master "${RANGE_END}")
+  START_REVISION="${START_REVISION/$RANGE_START/$NEW_RANGE_START}"
+
+  START_SHA="${START_REVISION}"
+
+elif ! git rev-parse origin/master >& /dev/null; then
+  git remote set-branches --add origin master
+  git fetch origin
+  START_SHA=$(git merge-base origin/master "${TRAVIS_COMMIT}")
```

Comment:
I don't understand the logic here.

If we're given a range, we handle it as above. If not given a range, why are we checking origin/master?

Reply:
To compute START_SHA when $TRAVIS_COMMIT_RANGE is empty.  See https://travis-ci.org/firebase/firebase-ios-sdk/jobs/536944398

Reply:
I understand the effect you're going for.

What I'm saying is that reading the code, it doesn't follow that if we're not processing a commit range that we're necessarily checking against `origin/master` and that `TRAVIS_COMMIT` would necessarily be set.

In particular, this could have been run by a user, and they pass a single commit that they want to check against, in which case it's irrelevant whether or not `origin/master` is available in the repo. Indeed the change you've made is harmful to this usage because `TRAVIS_COMMIT` is unset in that case.

All told I think we should cleanly separate the concerns of figuring out what we're checking from any repair of it. This change mixes these up in a way that's confusing and ultimately wrong for some use cases.

I recommend the following changes:
  * Change the travis configuration to not pass $TRAVIS_COMMIT_RANGE. This makes it unambiguous that travis passes arguments via environment variable.
  * Before parsing arguments, try to determine `START_REVISION` from travis environment variables
    * Something like `if [[ -n "${TRAVIS_COMMIT_RANGE:-}" ]]` would be the test to see if the variable is set and non-empty without violating `-o nounset`.
    * Check for both `TRAVIS_COMMIT_RANGE` and `TRAVIS_COMMIT`
    * In either case, all you do is assign to `START_REVISION`
  * Then down here:
    * Handle the range case essentially as we do, because travis can give busted ranges
    * If `START_REVISION` starts with `origin`, ensure that it's actually in the repo